presto-docs/src/main/sphinx/develop/delete-and-update.rst

```diff
@@ -0,0 +1,203 @@
+  collection of ``Slice`` fragments.  In response, the connector takes appropriate actions to complete the ``Delete`` operation.
+  Those actions might include committing the transaction, assuming the connector supports a transaction paradigm.
+
+``ConnectorMetadata`` ``UPDATE`` API
+====================================
+
+A connector implementing ``UPDATE`` must specify three ``ConnectorMetadata`` methods.
+
+* ``getUpdateRowIdColumnHandle``::
+
+   ColumnHandle getUpdateRowIdColumnHandle(
+        ConnectorSession session,
+        ConnectorTableHandle tableHandle,
+        List<ColumnHandle> updatedColumns)
+
+  The ``updatedColumns`` list contains column handles for all columns updated by the ``UPDATE`` operation in table column order.
+
+  The ColumnHandle returned by this method provides the "rowId" used by the connector to identify rows to be updated, as
+  well as any other fields of the row that the connector will need to complete the ``UPDATE`` operation.
+  For a JDBC connector, that rowId is usually the primary key for the table and no other fields are required.
+  For other connectors, the information needed to identify a row usually consists of multiple physical columns.
+  Moreover, some connectors may need the values of columns that are not updated to complete the ``UPDATE`` operation.
+  With the Hive connector and a Hive ACID table, for example, the ``UPDATE`` rowId consists of the three ORC ACID columns
+  that identify the row, plus the values of all the data columns not updated.
+
+* ``beginUpdate``::
+
+    ConnectorTableHandle beginUpdate(
+         ConnectorSession session,
+         ConnectorTableHandle tableHandle,
+         Map<String, ColumnHandle> columnHandles,
+         List<ColumnHandle> updatedColumns,
+         List<Integer> updateExpressionChannelNumbers)
+
+  As the last step in creating the ``UPDATE`` execution plan, the connector's ``beginUpdate()`` method is called,
+  passing arguments that define the ``UPDATE`` to the connector.  In addition to the ``session``
+  and ``tableHandle``, the arguments include the list of the updated columns and a list of channel numbers of
+  updated columns in the projected page created by the Presto engine and passed into ``UpdatablePageSource.updateRows()``.
+  All column lists are in table column order, as are the ``updateExpressionChannelNumbers``.  The size of
+  ``updateExpressionChannelNumbers`` is one greater than the number of updated columns - - the last element of the
+  list is the channel number of the block of rowIds for the page, constructed from the rowId column handle.
+
+  ``beginUpdate()`` performs any orchestration needed in the connector to start processing the ``UPDATE``.
+  This orchestration varies from connector to connector.  In the Hive ACID connector, for example, ``beginUpdate()``
+  starts the Hive Metastore transaction; checks that the updated table is transactional and that neither
```

Comment:
Might not be relevant, just thinking out loud. What would be the behavior if there is a concurrent Alter table which drops or renames the columns before the beginUpdate is invoked?

Reply:
It's a good question, and an area that is poorly tested.  If you could think of a way to test commit and rollback it would be a big contribution.  Here is the way I _think_ it works:

If the `ALTER TABLE `transaction committed before `beginUpdate() `was called, the `UPDATE` operation would fail because the `StatementAnalyzer` would fail to find a referenced column.  If the ALTER TABLE transaction was started but not committed by the time `beginUpdate()` was called, I think the call to `metastore.declareIntentionToWrite()` will cause a rollback of the `beginUpdate()` transaction when the `ALTER TABLE` transaction commits.  But I'm guessing, and don't have tests to prove that it works that way :(

Reply:
Assuming the metastore was a proper transactional system that implemented optimistic concurrency, the commit of the `UPDATE` operation would fail due to the conflict. That's how it will work for Iceberg, which properly supports transaction semantics.

Reply:
Makes sense that as long as we have OCC backed metastore, one of the conflicting transaction would be rejected. But, I was thinking can there be an Alter table request which makes the conflict not detectable because we rely on the names/order of the columns? e.g the scenarios listed here https://iceberg.apache.org/evolution/#correctness
```
    Formats that track columns by name can inadvertently un-delete a column if a name is reused, which violates #1.
    Formats that track columns by position cannot delete columns without changing the names that are used for each column, which violates #2.
```